File: Lab 9/knapsack.c
#include <stdio.h>

int main() {
    int profits[100], weights[100];
    int dp[101][101];
    int n, capacity;

    // User input for number of items
    printf("Enter the number of items: ");
    scanf("%d", &n);

    // User input for profits
    printf("Enter the profits of the items:\n");
    for (int i = 0; i < n; i++) {
        printf("Profit of item %d: ", i + 1);
        scanf("%d", &profits[i]);
    }

    // User input for weights
    printf("Enter the weights of the items:\n");
    for (int i = 0; i < n; i++) {
        printf("Weight of item %d: ", i + 1);
        scanf("%d", &weights[i]);
    }

    // User input for knapsack capacity
    printf("Enter the capacity of the knapsack: ");
    scanf("%d", &capacity);

     // Initialize dp array (Base case: dp[0][w] and dp[i][0] = 0)
    for (int i = 0; i <= n; i++) {
        for (int w = 0; w <= capacity; w++) {
            if (i == 0 || w == 0) {
                dp[i][w] = 0;
            } else if (weights[i - 1] <= w) {
                dp[i][w] = (profits[i - 1] + dp[i - 1][w - weights[i - 1]] > dp[i - 1][w]) ?
                            profits[i - 1] + dp[i - 1][w - weights[i - 1]] : dp[i - 1][w];
                 // Print decision-making process
                printf("Considering item %d (Profit: %d, Weight: %d) with capacity %d\n", 
                        i, profits[i - 1], weights[i - 1], w);
                printf("Include item %d: %d + dp[%d][%d] = %d\n", 
                        i, profits[i - 1], i - 1, w - weights[i - 1],dp[i][w]);
                printf("Exclude item %d: dp[%d][%d] = %d\n", 
                        i, i - 1, w,dp[i - 1][w]);
                printf("dp[%d][%d] = %d\n\n", i, w, dp[i][w]);
            } else {
                dp[i][w] = dp[i - 1][w];

                // Print decision-making process
                printf("Considering item %d (Profit: %d, Weight: %d) with capacity %d\n", 
                        i, profits[i - 1], weights[i - 1], w);
                printf("Item too heavy to include. dp[%d][%d] = dp[%d][%d] = %d\n\n", 
                        i, w, i - 1, w, dp[i][w]);
            }
        }
    }

    // Output the DP table for clarity
    printf("DP Table:\n");
    for (int i = 0; i <= n; i++) {
        for (int w = 0; w <= capacity; w++) {
            printf("%3d ", dp[i][w]);
        }
        printf("\n");
    }

    // Backtrack to find which items are included
    int w = capacity;
    printf("\nItems included to achieve maximum profit:\n");
    for (int i = n; i > 0; i--) {
        if (dp[i][w] != dp[i - 1][w]) {  // Item i is included
            printf("Item %d (Profit: %d, Weight: %d)\n", i, profits[i - 1], weights[i - 1]);
            w = w - weights[i - 1];  // Decrease the remaining capacity
        }
    }

    // Output the maximum profit
    printf("Maximum profit in the knapsack: %d\n", dp[n][capacity]);

    return 0;
}

File: Lab 9/knapsack.py
def knapsack():
    n = int(input("Enter the number of items: "))
    profits, weights = [], []

    # Input profits and weights
    for i in range(n):
        profits.append(int(input(f"Profit of item {i + 1}: ")))
    for j in range(n):    
        weights.append(int(input(f"Weight of item {j + 1}: ")))

    capacity = int(input("Enter the capacity of the knapsack: "))
    dp = [[0] * (capacity + 1) for _ in range(n + 1)]

    # Dynamic Programming table filling
    for i in range(1, n + 1):
        for w in range(1, capacity + 1):
            if weights[i - 1] <= w:
                include = profits[i - 1] + dp[i - 1][w - weights[i - 1]]
                exclude = dp[i - 1][w]
                dp[i][w] = max(include, exclude)

                # Print decision-making process
                print(f"Considering item {i} (Profit: {profits[i - 1]}, Weight: {weights[i - 1]}) with capacity {w}")
                print(f"Include item {i}: {profits[i - 1]} + dp[{i - 1}][{w - weights[i - 1]}] = {dp[i][w]}")
                print(f"Exclude item {i}: dp[{i - 1}][{w}] = {exclude}")
            else:
                dp[i][w] = dp[i - 1][w]
                print(f"Item {i} too heavy for capacity {w}, dp[{i}][{w}] = {dp[i][w]}")

            print(f"dp[{i}][{w}] = {dp[i][w]}\n")

    # Display DP table
    print("DP Table:")
    for row in dp:
        print(" ".join(f"{x:2d}" for x in row))

    # Backtrack to find selected items
    w = capacity
    print("\nItems included for maximum profit:")
    for i in range(n, 0, -1):
        if dp[i][w] != dp[i - 1][w]:  # Item is included
            print(f"Item {i} (Profit: {profits[i - 1]}, Weight: {weights[i - 1]})")
            w -= weights[i - 1]

    print(f"Maximum profit in the knapsack: {dp[n][capacity]}")

if __name__ == "__main__":
    knapsack()
